src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,189 @@
+//   Copyright 2024, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#include "LazyGroupBy.h"
+
+#include "engine/sparqlExpressions/LiteralExpression.h"
+#include "global/Constants.h"
+
+template <size_t NUM_GROUP_COLUMNS>
+LazyGroupBy<NUM_GROUP_COLUMNS>::LazyGroupBy(
+    LocalVocab& localVocab,
+    std::vector<GroupBy::HashMapAliasInformation> aggregateAliases,
+    const std::vector<size_t>& groupByCols,
+    const ad_utility::AllocatorWithLimit<Id>& allocator)
+    : localVocab_{localVocab},
+      aggregateAliases_{std::move(aggregateAliases)},
+      groupByCols_{groupByCols},
+      aggregationData_{allocator, aggregateAliases_, NUM_GROUP_COLUMNS} {
+  for (const auto& alias : aggregateAliases_) {
+    for (const auto& aggregateInfo : alias.aggregateInfo_) {
+      std::visit(
+          [&aggregateInfo]<SupportedAggregates T>(T& arg) {
+            if constexpr (std::same_as<typename T::value_type,
+                                       GroupConcatAggregationData>) {
+              arg.resize(1,
+                         GroupConcatAggregationData{
+                             aggregateInfo.aggregateType_.separator_.value()});
+            } else {
+              arg.resize(1);
+            }
+          },
+          aggregationData_.getAggregationDataVariant(
+              aggregateInfo.aggregateDataIndex_));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS>
+void LazyGroupBy<NUM_GROUP_COLUMNS>::resetAggregationData() {
+  for (const auto& alias : aggregateAliases_) {
+    for (const auto& aggregateInfo : alias.aggregateInfo_) {
+      std::visit([]<SupportedAggregates T>(T& arg) { arg.at(0).reset(); },
+                 aggregationData_.getAggregationDataVariant(
+                     aggregateInfo.aggregateDataIndex_));
```

Comment:
I have seen this rather complicated `nested for, then visit` at least twice,
I think some helpers like
`forEachAggregate` or `forEachAggregationVector` (names are stupid) can make the code much more readable.